README.md
# Google Firebase Cloud Messaging Cordova Push Plugin
> Extremely easy plug&play push notification plugin for Cordova applications with Google Firebase FCM.

Forked from https://github.com/andrehtissot/cordova-plugin-fcm-with-dependecy-updated which is no longer maintained.

[How it works](#how-it-works) | [Installation](#installation) | [Push Payload Configuration](#push-payload-configuration) |  [Features](#features)

## How it works
Send a push notification to a single device or topic.

- Application is in foreground:

  - The notification data is received in the JavaScript callback without notification bar message (this is the normal behavior of mobile push notifications).
  - For Android, to show the notification received on the foreground, it's recommended to use [cordova-plugin-local-notification](https://github.com/katzer/cordova-plugin-local-notifications#readme) as it provides many presentation and interaction features.

- Application is in background or closed:

  - The device displays the notification message in the device notification bar.
  - If the user taps the notification, the application comes to foreground and the notification data is received in the JavaScript callback.
  - If the user does not tap the notification but opens the application, nothing happens until the notification is tapped.
  
### Push Notifications on iOS

On Android, push notifications can be tested from emulators freely.

Unfortunately, Apple is not as nice to work with, requiring:
- The running device to be a real device, no simulators allowed;
- Application has require the `UIBackgroundModes=[remote-notification]` permission (automatically configured by this plugin);
- The user running the application has to manually allow the application to receive push notifications;
- The application must be build with credentials created from a paid account (or team) that is allowed to receive push notifications;
- The build installed has to have come from either Apple Store or TestFlight; Or, be build with a special certificate (https://customersupport.doubledutch.me/hc/en-us/articles/229495568-iOS-How-to-Create-a-Push-Notification-Certificate)

## Installation

Make sure you have ‘google-services.json’ for Android and/or ‘GoogleService-Info.plist’ for iOS in your Cordova project root folder.

#### Preferences

|Preference| Default Value       |Description|
|---|---------------------|---|
|ANDROID_DEFAULT_NOTIFICATION_ICON| @mipmap/ic_launcher |Default notification icon.|
|ANDROID_FCM_VERSION| 23.0.8              |Native Firebase Message SDK version.<br>:warning: Replaced by BoM versioning on Gradle >= 3.4.|
|ANDROID_FIREBASE_BOM_VERSION| 29.0.1              |[Firebase BoM](https://firebase.google.com/docs/android/learn-more#bom) version.|
|ANDROID_GOOGLE_SERVICES_VERSION| 4.3.4               |Native Google Services SDK version.|
|ANDROID_GRADLE_TOOLS_VERSION| 4.1.0               |Gradle tools version.|
|IOS_FIREBASE_MESSAGING_VERSION| ~> 7.4.0            |Native Firebase Message SDK version|

#### Cordova

Default:

```sh
npm uninstall @ionic-native/fcm # Ionic support is included and conflicts with @ionic-native's implementation.
npm install cordova-plugin-fcm-with-dependecy-updated@git+https://github.com/4sh/cordova-plugin-fcm-with-dependecy-updated.git#7.10.0
```

Complete:

```sh
npm uninstall @ionic-native/fcm # Ionic support is included and conflicts with @ionic-native's implementation.
npm install cordova-plugin-fcm-with-dependecy-updated@git+https://github.com/4sh/cordova-plugin-fcm-with-dependecy-updated.git#7.10.0 \
  --variable ANDROID_DEFAULT_NOTIFICATION_ICON="@mipmap/ic_launcher" \
  --variable ANDROID_FIREBASE_BOM_VERSION="26.0.0" \
  --variable ANDROID_GOOGLE_SERVICES_VERSION="4.3.4" \
  --variable ANDROID_GRADLE_TOOLS_VERSION="4.1.0" \
  --variable IOS_FIREBASE_MESSAGING_VERSION="~> 7.4.0"
```

## Push Payload Configuration

Besides common FCM configuration (https://firebase.google.com/docs/cloud-messaging/ios/certs), the Push payload should contain "notification" and "data" keys and "click_action" equals to "FCM_PLUGIN_ACTIVITY" within "notification".

Structure expected:
```js
{
  ...,
  "notification": {
    ...
  },
  "data": {
    ...
  },
  "android": {
    "notification": {
      "click_action": "FCM_PLUGIN_ACTIVITY"
    }
  },
  ...,
}
```

Example:
```json
{
  "token": "[FCM token]",
  "notification":{
    "title":"Notification title",
    "body":"Notification body",
    "sound":"default",
  },
  "data":{
    "param1":"value1",
    "param2":"value2"
  },
  "android": {
    "notification": {
      "icon":"fcm_push_icon",
      "click_action": "FCM_PLUGIN_ACTIVITY"
    }
  }
}
```


## Features

- [As its own](#as-its-own)
  - [FCM.clearAllNotifications()](#fcmclearallnotifications)
  - [FCM.createNotificationChannel()](#fcmcreatenotificationchannel)
  - [FCM.deleteInstanceId()](#fcmdeleteinstanceid)
  - [FCM.getAPNSToken()](#fcmgetapnstoken)
  - [FCM.getInitialPushPayload()](#fcmgetinitialpushpayload)
  - [FCM.getToken()](#fcmgettoken)
  - [FCM.hasPermission()](#fcmhaspermission)
  - [FCM.onNotification()](#fcmonnotification)
  - [FCM.onTokenRefresh()](#fcmontokenrefresh)
  - [**FCM.requestPushPermission()**](#fcmrequestpushpermission)
  - [FCM.subscribeToTopic()](#fcmsubscribetotopic)
  - [FCM.unsubscribeFromTopic()](#fcmunsubscribefromtopic)
  - [FCM.eventTarget](#fcmeventtarget)
- [**With Ionic**](#with-ionic)
  - [FCM.onNotification()](#fcmonnotification-1)
  - [FCM.onTokenRefresh()](#fcmontokenrefresh-1)

#### As its own

The JS functions are now as written bellow and do require Promise support. Which, for Android API 19 support, it can be fulfilled by a polyfill.

##### FCM.clearAllNotifications()

Removes existing push notifications from the notifications center.
```typescript
await FCM.clearAllNotifications();
```

##### FCM.createNotificationChannel()

For Android, some notification properties are only defined programmatically.
Channel can define the default behavior for notifications on Android 8.0+.
Once a channel is created, it stays unchangeable until the user uninstalls the app.
```typescript
await FCM.createNotificationChannel({
  id: "urgent_alert", // required
  name: "Urgent Alert", // required
  description: "Very urgent message alert",
  importance: "high", // https://developer.android.com/guide/topics/ui/notifiers/notifications#importance
  visibility: "public", // https://developer.android.com/training/notify-user/build-notification#lockscreenNotification
  sound: "alert_sound", // In the "alert_sound" example, the file should located as resources/raw/alert_sound.mp3
  lights: true, // enable lights for notifications
  vibration: true // enable vibration for notifications
});
```

##### FCM.deleteInstanceId()

Deletes the InstanceId, revoking all tokens.
```typescript
await FCM.deleteInstanceId();
```

##### FCM.getAPNSToken()

Gets iOS device's current APNS token.
```typescript
const apnsToken: string = await FCM.getAPNSToken();
```

##### FCM.getInitialPushPayload()

Retrieves the message that, on tap, opened the app. And `null`, if the app was open normally.
```typescript
const pushPayload: object = await FCM.getInitialPushPayload()
```

##### FCM.getToken()

Gets device's current registration id.
```typescript
const fcmToken: string = await FCM.getToken()
```

##### FCM.hasPermission()

On iOS, returns `true` if runtime permission for remote notifications is granted and enabled in Settings.
On Android, returns `true` if global remote notifications are enabled in the device settings and, from Android 13+, runtime permission for remote notifications is granted.

```typescript
const doesIt: boolean = await FCM.hasPermission()
```

##### FCM.onNotification()

Callback firing when receiving new notifications. It serves as a shortcut to listen to eventTarget's "notification" event.
```typescript
const disposable = FCM.onNotification((payload: object) => {
  // ...
})
// ...
disposable.dispose() // To remove listener
```

:warning: If the subscription to notification events happens after the notification has been fired, it'll be lost. As it is expected that you'd not always be able to catch the notification payload that the opened the app, the `FCM.getInitialPushPayload()` method was introduced.

##### FCM.onTokenRefresh()

Callback firing when receiving a new Firebase token. It serves as a shortcut to listen to eventTarget's "tokenRefresh" event.
```typescript
const disposable = FCM.onTokenRefresh((fcmToken: string) => {
  // ...
})
// ...
disposable.dispose() // To remove listener
```

##### FCM.requestPushPermission()

Grant run-time permission to receive push notifications (will trigger user permission dialog).
iOS & Android 13+ (Android <= 12 does not display any system dialog and immediately returns the value of `hasPermission`).

On Android, the `POST_NOTIFICATIONS` permission has been added to the `AndroidManifest.xml`.

```typescript
const wasPermissionGiven: boolean = await FCM.requestPushPermission({
  ios9Support: {
    timeout: 10,  // How long it will wait for a decision from the user before returning `false`
    interval: 0.3 // How long between each permission verification
  }
})
```

:warning: Without this request, the application won't receive any notification on iOS or Android 13+ (notifications are turned off by default)!

On iOS, the user will only have its permission required once, after that time, this call will only return if the permission was given that time. 
On Android, your app has complete control over when the permission dialog is displayed. Use this opportunity to explain to users why the app needs this permission, encouraging them to grant it. See [best practices](https://developer.android.com/develop/ui/views/notifications/notification-permission#wait-to-show-prompt).

##### FCM.subscribeToTopic()

Subscribes you to a [topic](https://firebase.google.com/docs/notifications/android/console-topics).
```typescript
const topic: string
// ...
await FCM.subscribeToTopic(topic)
```

##### FCM.unsubscribeFromTopic()

Unsubscribes you from a [topic](https://firebase.google.com/docs/notifications/android/console-topics).
```typescript
const topic: string
// ...
await FCM.unsubscribeFromTopic(topic)
```

##### FCM.eventTarget

EventTarget object for native-sourced custom events. Useful for more advanced listening handling.
```typescript
const listener = (data) => {
  const payload = data.detail
  // ...
}
FCM.eventTarget.addEventListener("notification", listener, false);
// ...
FCM.eventTarget.removeEventListener("notification", listener, false);
```

#### With Ionic

Ionic support was implemented as part of this plugin to allow users to have access to newer features with the type support.

- Ionic v5:
```typescript
import {FCM} from 'cordova-plugin-fcm-with-dependecy-updated/ionic';

// ...
private listenToInitialPushNotification(): Subscription {
  return from(FCM.getInitialPushPayload()).pipe(
          filter(data => data && data.wasTapped),
  ).subscribe(data => console.log(data));
}

```

It brings the same behavior as the native implementation, except for `FCM.onNotification()` and `FCM.onTokenRefresh()`, which gain rxjs' Observable support.

To avoid confusion, it's suggested to also remove the redundant `@ionic-native/fcm` package.

##### FCM.onNotification()

Event firing when receiving new notifications.
```typescript
FCM.onNotification().subscribe((payload: object) => {
  // ...
});
```

##### FCM.onTokenRefresh()

Event firing when receiving a new Firebase token.
```typescript
FCM.onTokenRefresh().subscribe((token: string) => {
  // ...
});
```



CHANGELOG.md
# Changelog

## Version 7

### Version 7.10.1 (08/11/2023)

- Remove post install script (for installing ionic native dependencies) causing issue with recent Node.js versions 

### Version 7.10.0 (11/07/2023)

- Support Android SDK level 33 (Android 13): user permission is now required to start receiving push notifications from Android 13+. Calling `FCM.requestPushPermission()` will now display a push permission dialog in Android 13 or above.

### Version 7.9.0 (17/11/2023)

- Support Android SDK level 31 (Android 12)

### Version 7.8.0 (03/02/2020)

- IOS_FIREBASE_MESSAGING_VERSION upgraded to 7.4.0;
- FCM.hasPermission now supports Android;
- On iOS, getToken will now wait until fcm token is defined to return it;
- FCM.getInitialPushPayload now uses UTF8 instead of ISOLatin;

### Version 7.7.0 (11/12/2020)

Downgraded Cordova and Cordova-Android minimal required versions.

### Version 7.6.1 (11/12/2020)

Fixed auto install of ionic-specific dependencies on Windows

### Version 7.6.0 (08/11/2020)

Added IOS_FIREBASE_MESSAGING_VERSION plugin variable to force a fixed Firebase/Messaging pod version.

### Version 7.5.0 (08/11/2020)

For pure Cordova projects, ionic dependencies are skipped for "ionic", "ionic/ngx" and "ionic/v4".

### Version 7.4.0 (01/11/2020)

Upgraded Android and Node depenencies.

### Version 7.3.1 (06/09/2020)

Android: Avoided setting initialPushPayload from a non-tapped notification.

### Version 7.3.0 (30/08/2020)

Removed optional Firebase/Analytics iOS dependency.

### Version 7.2.0 (23/08/2020)

Added deleteInstanceId to allow user's "resetting" Firebase session (https://firebase.google.com/support/privacy/manage-iids#delete_an_instance_id).

### Version 7.1.1 (16/08/2020)

iOS: Values sent as notification on push payload should never overwrite values sent as data.

Android: tapping action small refactor.

Ionic: Removed tslib dependency due to reported incompatibility.

### Version 7.1.0 (16/08/2020)

For iOS: Added `title`, `subtitle`, `body` and `badge` to the data given to JS.

This data is coming from notification part of the push payload, instead of only from data.

Android has been responding this way for a long time with `title` and `body`.

### Version 7.0.10 (16/08/2020)

Defined minimal version of "cordova" package supported as version 9. Due to issues of supporting lower versions and "cordova-ios@5+" without deprecated configuration.

Removed explicit definition of "Firebase/Analytics" and "Firebase/Messaging" due to lack of matching “>=” definitions between plugins.

### Version 7.0.9 (21/07/2020)

getInitialPushPayload: Conversion of NSData* into NSDictonary* fix -- Thanks to @medeirosrafael for debugging and fixing it!

### Version 7.0.8 (20/07/2020)

Avoided execution of install_ionic_dependencies.bat after the app is installed;

Removed old framework dependencies -- Thanks for @QuentinFarizon, for pointing it out!

### Version 7.0.7 (17/07/2020)

Removed auto-cdnfy due to service issues;

Set AndroidXEnabled for cordova-android-9.0.0 (https://cordova.apache.org/announcements/2020/06/29/cordova-android-9.0.0.html).

### Version 7.0.6 (14/07/2020)

Simplified ionic/ngx/FCM.d.ts imports

### Version 7.0.5 (13/07/2020)

Renamed extension scripts/install_ionic_dependencies.sh to .bat, to have it running on windows.

### Version 7.0.4 (13/07/2020)

Simplified ionic/FCM.js and ionic/ngx/FCM.js files to ease building with them.

Improved scripts/install_ionic_dependencies.sh windows support.

### Version 7.0.3 (07/07/2020)

Simplified ionic/v4/FCM.js file by setting the FCM function in the global context.

### Version 7.0.2 (01/07/2020)

Simplified .d.ts files by removing the new "type" from imports and exports.

### Version 7.0.1 (28/06/2020)

Replaced native to JS context messaging, from JS injection to event subscription.

### Version 7.0.0 (27/06/2020)

Breaking update released. Please pay atention to the changes in plugin preferences and API methods.

## Version 6

### Version 6.4.0 (21/05/2020)

The `FCMPlugin.requestPushPermissionIOS` function now, not only triggers the request alert, but also returns, as boolean, if the permission was given.

```javascript
FCMPlugin.requestPushPermissionIOS(
  function(wasPermissionGiven) {
    console.info("Was permission given: "+wasPermissionGiven);
  },
  function(error) {
    console.error(error);
  },
  ios9Support: {
    timeout: 10,  // How long it will wait for a decision from the user before returning `false`
    interval: 0.3 // How long between each permission verification
  }
);
```

Note:
On iOS 9, there is no way to know if the user has denied the permissions or he has not yet decided.
For this reason, specifically for iOS 9, after presenting the alert, a timed loop waits until it knows that the user has either given the permissions or that the time has expired.
On iOS 10+, the return is given as soon as the user has selected, ignoring these options.

### Version 6.3.0 (27/04/2020)

FCMPlugin.createNotificationChannelAndroid improved, now accepting three other options: "sound", "lights" and "vibration", like in:
```javascript
FCMPlugin.createNotificationChannelAndroid({
  id: "urgent_alert", // required
  name: "Urgent Alert", // required
  description: "Very urgent message alert",
  importance: "high", // https://developer.android.com/guide/topics/ui/notifiers/notifications#importance
  visibility: "public", // https://developer.android.com/training/notify-user/build-notification#lockscreenNotification
  sound: "alert_sound", // In the "alert_sound" example, the file should located as resources/raw/alert_sound.mp3
  lights: true, // enable lights for notifications
  vibration: true // enable vibration for notifications
});
```

### Version 6.2.0 (26/04/2020)

IOS 9 support reintroduced.

### Version 6.1.0 (24/04/2020)


src/android/com/gae/scaffolder/plugin/FCMPlugin.java
package com.gae.scaffolder.plugin;

import androidx.core.app.NotificationManagerCompat;
import android.app.NotificationManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.Log;

import com.gae.scaffolder.plugin.interfaces.*;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.iid.InstanceIdResult;
import com.google.firebase.messaging.FirebaseMessaging;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Map;

public class FCMPlugin extends CordovaPlugin {
    public static String notificationEventName = "notification";
    public static String tokenRefreshEventName = "tokenRefresh";
    public static Map<String, Object> initialPushPayload;
    public static final String TAG = "FCMPlugin";
    private static FCMPlugin instance;
    protected Context context;
    protected static CallbackContext jsEventBridgeCallbackContext;
    protected static final String POST_NOTIFICATIONS = "POST_NOTIFICATIONS";
    protected static final int POST_NOTIFICATIONS_PERMISSION_REQUEST_ID = 1;

    private static CallbackContext postNotificationPermissionRequestCallbackContext;

    public FCMPlugin() {
    }

    public FCMPlugin(Context context) {
        this.context = context;
    }

    public static synchronized FCMPlugin getInstance(Context context) {
        if (instance == null) {
            instance = new FCMPlugin(context);
            instance = getPlugin(instance);
        }

        return instance;
    }

    public static synchronized FCMPlugin getInstance() {
        if (instance == null) {
            instance = new FCMPlugin();
            instance = getPlugin(instance);
        }

        return instance;
    }

    public static FCMPlugin getPlugin(FCMPlugin plugin) {
        if (plugin.webView != null) {
            instance = (FCMPlugin) plugin.webView.getPluginManager().getPlugin(FCMPlugin.class.getName());
        } else {
            plugin.initialize(null, null);
            instance = plugin;
        }

        return instance;
    }

    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        Log.d(TAG, "==> FCMPlugin initialize");

        FirebaseMessaging.getInstance().subscribeToTopic("android");
        FirebaseMessaging.getInstance().subscribeToTopic("all");
    }

    public boolean execute(final String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        Log.d(TAG, "==> FCMPlugin execute: " + action);

        try {
            if (action.equals("ready")) {
                callbackContext.success();
            } else if (action.equals("startJsEventBridge")) {
                this.jsEventBridgeCallbackContext = callbackContext;
            } else if (action.equals("getToken")) {
                cordova.getActivity().runOnUiThread(new Runnable() {
                    public void run() {
                        getToken(callbackContext);
                    }
                });
            } else if (action.equals("getInitialPushPayload")) {
                cordova.getActivity().runOnUiThread(new Runnable() {
                    public void run() {
                        getInitialPushPayload(callbackContext);
                    }
                });
            } else if (action.equals("subscribeToTopic")) {
                cordova.getThreadPool().execute(new Runnable() {
                    public void run() {
                        try {
                            FirebaseMessaging.getInstance().subscribeToTopic(args.getString(0));
                            callbackContext.success();
                        } catch (Exception e) {
                            callbackContext.error(e.getMessage());
                        }
                    }
                });
            } else if (action.equals("unsubscribeFromTopic")) {
                cordova.getThreadPool().execute(new Runnable() {
                    public void run() {
                        try {
                            FirebaseMessaging.getInstance().unsubscribeFromTopic(args.getString(0));
                            callbackContext.success();
                        } catch (Exception e) {
                            callbackContext.error(e.getMessage());
                        }
                    }
                });
            } else if (action.equals("clearAllNotifications")) {
                cordova.getThreadPool().execute(new Runnable() {
                    public void run() {
                        try {
                            Context context = cordova.getActivity();
                            NotificationManager nm = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
                            nm.cancelAll();
                            callbackContext.success();
                        } catch (Exception e) {
                            callbackContext.error(e.getMessage());
                        }
                    }
                });
            } else if (action.equals("createNotificationChannel")) {
                cordova.getActivity().runOnUiThread(new Runnable() {
                    public void run() {
                        new FCMPluginChannelCreator(getContext()).createNotificationChannel(callbackContext, args);
                    }
                });
            } else if (action.equals("deleteInstanceId")) {
                this.deleteInstanceId(callbackContext);
            } else if (action.equals("hasPermission")) {
                this.hasPermission(callbackContext);
            } else if (action.equals("requestPushPermission")) {
                this.grantPermission(callbackContext);
            } else {
                callbackContext.error("Method not found");
                return false;
            }
        } catch (Exception e) {
            Log.d(TAG, "ERROR: onPluginAction: " + e.toString());
            callbackContext.error(e.getMessage());
            return false;
        }

        return true;
    }

    public void getInitialPushPayload(CallbackContext callback) {
        if (initialPushPayload == null) {
            Log.d(TAG, "getInitialPushPayload: null");
            callback.success((String) null);
            return;
        }
        Log.d(TAG, "getInitialPushPayload");
        try {
            JSONObject jo = new JSONObject();
            for (String key : initialPushPayload.keySet()) {
                jo.put(key, initialPushPayload.get(key));
                Log.d(TAG, "\tinitialPushPayload: " + key + " => " + initialPushPayload.get(key));
            }
            callback.success(jo);
        } catch (Exception error) {
            try {
                callback.error(exceptionToJson(error));
            } catch (JSONException jsonErr) {
                Log.e(TAG, "Error when parsing json", jsonErr);
            }
        }
    }

    public void getToken(final TokenListeners<String, JSONObject> callback) {
        try {
            FirebaseMessaging.getInstance().getToken().addOnCompleteListener(new OnCompleteListener<String>() {
                @Override
                public void onComplete(Task<String> task) {
                    if (!task.isSuccessful()) {
                        Log.w(TAG, "getInstanceId failed", task.getException());
                        try {
                            callback.error(exceptionToJson(task.getException()));
                        } catch (JSONException jsonErr) {
                            Log.e(TAG, "Error when parsing json", jsonErr);
                        }
                        return;
                    }

                    // Get new Instance ID token
                    String newToken = task.getResult();

                    Log.i(TAG, "\tToken: " + newToken);
                    callback.success(newToken);
                }
            });

            FirebaseMessaging.getInstance().getToken().addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(final Exception e) {
                    try {
                        Log.e(TAG, "Error retrieving token: ", e);
                        callback.error(exceptionToJson(e));
                    } catch (JSONException jsonErr) {
                        Log.e(TAG, "Error when parsing json", jsonErr);
                    }
                }
            });
        } catch (Exception e) {
            Log.w(TAG, "\tError retrieving token", e);
            try {
                callback.error(exceptionToJson(e));
            } catch (JSONException je) {
            }
        }
    }

    private void deleteInstanceId(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    FirebaseMessaging.getInstance().deleteToken();
                    callbackContext.success();
                } catch (Exception e) {
                    callbackContext.error(e.toString());
                }
            }
        });
    }

    private void hasPermission(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    boolean hasRuntimePermission;

                    if (Build.VERSION.SDK_INT >= 33) { // Android 13+
                        hasRuntimePermission = hasRuntimePermission(POST_NOTIFICATIONS);
                    } else {
                        hasRuntimePermission = true;
                    }

                    callbackContext.success(areNotificationsEnabled() && hasRuntimePermission ? 1 : 0);
                } catch (Exception e) {
                    callbackContext.error(e.toString());
                }
            }
        });
    }

    private boolean areNotificationsEnabled() {
        NotificationManagerCompat notificationManagerCompat =
                NotificationManagerCompat.from(cordova.getActivity().getApplicationContext());

        return notificationManagerCompat.areNotificationsEnabled();
    }

    private void grantPermission(final CallbackContext callbackContext) {
        CordovaPlugin plugin = this;
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    if (Build.VERSION.SDK_INT >= 33) { // Android 13+
                        boolean hasRuntimePermission = hasRuntimePermission(POST_NOTIFICATIONS);
                        if (!hasRuntimePermission) {
                            String[] permissions = new String[]{qualifyPermission(POST_NOTIFICATIONS)};
                            postNotificationPermissionRequestCallbackContext = callbackContext;
                            requestPermissions(plugin, POST_NOTIFICATIONS_PERMISSION_REQUEST_ID, permissions);
                            sendEmptyPluginResultAndKeepCallback(callbackContext);
                        }
                    } else {
                        callbackContext.success(areNotificationsEnabled() ? 1 : 0);
                    }

                } catch (Exception e) {
                    callbackContext.error(e.toString());
                }
            }
        });
    }

    protected void sendEmptyPluginResultAndKeepCallback(CallbackContext callbackContext) {
        PluginResult pluginresult = new PluginResult(PluginResult.Status.NO_RESULT);
        pluginresult.setKeepCallback(true);
        callbackContext.sendPluginResult(pluginresult);
    }

    protected String qualifyPermission(String permission) {
        if (permission.startsWith("android.permission.")) {
            return permission;
        } else {
            return "android.permission." + permission;
        }
    }

    protected boolean hasRuntimePermission(String permission) throws Exception {
        boolean hasRuntimePermission = true;
        String qualifiedPermission = qualifyPermission(permission);
        java.lang.reflect.Method method = null;
        try {
            method = cordova.getClass().getMethod("hasPermission", qualifiedPermission.getClass());
            Boolean bool = (Boolean) method.invoke(cordova, qualifiedPermission);
            hasRuntimePermission = bool.booleanValue();
        } catch (NoSuchMethodException e) {
            Log.w(TAG, "Cordova v" + CordovaWebView.CORDOVA_VERSION + " does not support runtime permissions so defaulting to GRANTED for " + permission);
        }
        return hasRuntimePermission;
    }

    protected void requestPermissions(CordovaPlugin plugin, int requestCode, String[] permissions) throws Exception {
        try {
            java.lang.reflect.Method method = cordova.getClass().getMethod("requestPermissions", org.apache.cordova.CordovaPlugin.class, int.class, java.lang.String[].class);
            method.invoke(cordova, plugin, requestCode, permissions);
        } catch (NoSuchMethodException e) {
            throw new Exception("requestPermissions() method not found in CordovaInterface implementation of Cordova v" + CordovaWebView.CORDOVA_VERSION);
        }
    }

    /************
     * Overrides
     ***********/

    /**
     * then updates the list of status based on the grantResults before passing the result back via the context.
     *
     * @param requestCode  - ID that was used when requesting permissions
     * @param permissions  - list of permissions that were requested
     * @param grantResults - list of flags indicating if above permissions were granted or denied
     */
    public void onRequestPermissionResult(int requestCode, String[] permissions, int[] grantResults) throws JSONException {
        String sRequestId = String.valueOf(requestCode);
        Log.v(TAG, "Received result for permissions request id=" + sRequestId);
        try {
            if (postNotificationPermissionRequestCallbackContext == null) {
                Log.e(TAG, "No callback context found for permissions request id=" + sRequestId);
                return;
            }

            boolean postNotificationPermissionGranted = false;
            for (int i = 0, len = permissions.length; i < len; i++) {
                String androidPermission = permissions[i];

                if (androidPermission.equals(qualifyPermission(POST_NOTIFICATIONS))) {
                    postNotificationPermissionGranted = grantResults[i] == PackageManager.PERMISSION_GRANTED;
                }
            }

            postNotificationPermissionRequestCallbackContext.success(postNotificationPermissionGranted ? 1 : 0);
            postNotificationPermissionRequestCallbackContext = null;

        } catch (Exception e) {
            if (postNotificationPermissionRequestCallbackContext != null) {
                postNotificationPermissionRequestCallbackContext.error(e.toString());
            } else {
                Log.e(TAG, "onRequestPermissionResult error " + e.toString());
            }
        }
    }

    private JSONObject exceptionToJson(final Exception exception) throws JSONException {
        return new JSONObject() {
            {
                put("message", exception.getMessage());
                put("cause", exception.getClass().getName());
                put("stacktrace", exception.getStackTrace().toString());
            }
        };
    }

    public void getToken(final CallbackContext callbackContext) {
        this.getToken(new TokenListeners<String, JSONObject>() {
            @Override
            public void success(String message) {
                callbackContext.success(message);
            }

            @Override
            public void error(JSONObject message) {
                callbackContext.error(message);
            }
        });
    }

    private static void dispatchJSEvent(String eventName, String stringifiedJSONValue) throws Exception {
        String jsEventData = "[\"" + eventName + "\"," + stringifiedJSONValue + "]";
        PluginResult dataResult = new PluginResult(PluginResult.Status.OK, jsEventData);
        dataResult.setKeepCallback(true);
        if (FCMPlugin.jsEventBridgeCallbackContext == null) {
            Log.d(TAG, "\tUnable to send event due to unreachable bridge context");
            return;
        }
        FCMPlugin.jsEventBridgeCallbackContext.sendPluginResult(dataResult);
        Log.d(TAG, "\tSent event: " + eventName + " with " + stringifiedJSONValue);
    }

    public static void setInitialPushPayload(Map<String, Object> payload) {
        if (initialPushPayload == null) {
            initialPushPayload = payload;
        }
    }

    public static void sendPushPayload(Map<String, Object> payload) {
        Log.d(TAG, "==> FCMPlugin sendPushPayload");
        try {
            JSONObject jo = new JSONObject();
            for (String key : payload.keySet()) {
                jo.put(key, payload.get(key));
                Log.d(TAG, "\tpayload: " + key + " => " + payload.get(key));
            }
            FCMPlugin.dispatchJSEvent(notificationEventName, jo.toString());
        } catch (Exception e) {
            Log.d(TAG, "\tERROR sendPushPayload: " + e.toString());
        }
    }

    public static void sendTokenRefresh(String token) {
        Log.d(TAG, "==> FCMPlugin sendTokenRefresh");
        try {
            FCMPlugin.dispatchJSEvent(tokenRefreshEventName, "\"" + token + "\"");
        } catch (Exception e) {
            Log.d(TAG, "\tERROR sendTokenRefresh: " + e.toString());
        }
    }

    @Override
    public void onDestroy() {
        initialPushPayload = null;
        jsEventBridgeCallbackContext = null;
    }

    protected Context getContext() {
        context = cordova != null ? cordova.getActivity().getBaseContext() : context;
        if (context == null) {
            throw new RuntimeException("The Android Context is required. Verify if the 'activity' or 'context' are passed by constructor");
        }

        return context;
    }
}
